miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/permission/MiniAppCustomPermissionCache.kt

```diff
@@ -3,20 +3,52 @@ package com.rakuten.tech.mobile.miniapp.permission
 import android.content.Context
 import android.content.SharedPreferences
 import androidx.annotation.VisibleForTesting
+import androidx.security.crypto.EncryptedSharedPreferences
+import androidx.security.crypto.MasterKeys
 import com.google.gson.Gson
 import com.google.gson.reflect.TypeToken
+import com.rakuten.tech.mobile.miniapp.MiniAppVerificationException
 import java.lang.Exception
 
 /**
  * A caching class to read and store the grant results of custom permissions per MiniApp
  * using [SharedPreferences].
  */
 @Suppress("TooGenericExceptionCaught", "SwallowedException", "LongMethod")
-internal class MiniAppCustomPermissionCache(context: Context) {
-    private val prefs: SharedPreferences = context.getSharedPreferences(
-        "com.rakuten.tech.mobile.miniapp.custom.permissions.cache", Context.MODE_PRIVATE
+internal class MiniAppCustomPermissionCache constructor(
+    private val nonEncryptedPreferences: SharedPreferences,
+    private val prefs: SharedPreferences
+) {
+
+    constructor(context: Context) : this(
+        nonEncryptedPreferences = context.getSharedPreferences(
+            "com.rakuten.tech.mobile.miniapp.custom.permissions.cache", Context.MODE_PRIVATE
+        ),
+        prefs = initEncryptedSharedPreference(context)
     )
 
+    init {
+        migrateToEncryptedPref()
+    }
+
+    private fun migrateToEncryptedPref() {
+        if (nonEncryptedPreferences.all.isNotEmpty()) {
+            nonEncryptedPreferences.copyTo(prefs)
+            nonEncryptedPreferences.edit().clear().apply()
+        }
+    }
+
+    private fun SharedPreferences.copyTo(dest: SharedPreferences) = with(dest.edit()) {
+        for ((key, value) in all) {
+            val v = value ?: continue
+            when (v) {
+                is String -> putString(key, v)
+                else -> error("Unknown type: $v")
```

Comment:
I think that the Host App isn't able to catch this anywhere, so this Exception would crash the App, right? In what cases could this occur?

It seems like maybe we should just log the error instead of throwing an exception. What do you think?

Reply:
I have updated this area.